apm-lambda-extension/main.go

```diff
@@ -125,68 +117,20 @@ func main() {
 			extension.Log.Debug("Received event.")
 			extension.Log.Debugf("%v", extension.PrettyPrint(event))
 
-			// Make a channel for signaling that we received the agent flushed signal
-			extension.AgentDoneSignal = make(chan struct{})
-			// Make a channel for signaling that we received the runtimeDone logs API event
-			runtimeDoneSignal := make(chan struct{})
 			// Make a channel for signaling that the function invocation is complete
 			funcDone := make(chan struct{})
 
 			// A shutdown event indicates the execution environment is shutting down.
 			// This is usually due to inactivity.
 			if event.EventType == extension.Shutdown {
-				extension.ProcessShutdown()
+				extension.ProcessShutdown(agentDataServer, logsTransport.Server)
 				cancel()
 				return
 			}
 
-			// Receive agent data as it comes in and post it to the APM server.
-			// Stop checking for, and sending agent data when the function invocation
-			// has completed, signaled via a channel.
 			backgroundDataSendWg.Add(1)
-			go func() {
-				defer backgroundDataSendWg.Done()
-				if !extension.IsTransportStatusHealthyOrPending() {
-					return
-				}
-				for {
-					select {
-					case <-funcDone:
-						extension.Log.Debug("Received signal that function has completed, not processing any more agent data")
-						return
-					case agentData := <-agentDataChannel:
-						if err := extension.PostToApmServer(client, agentData, config, ctx); err != nil {
-							extension.Log.Errorf("Error sending to APM server, skipping: %v", err)
-							return
-						}
-					}
-				}
-			}()
-
-			// Receive Logs API events
-			// Send to the runtimeDoneSignal channel to signal when a runtimeDone event is received
-			go func() {
-				for {
-					select {
-					case <-funcDone:
-						extension.Log.Debug("Received signal that function has completed, not processing any more log events")
-						return
-					case logEvent := <-logsChannel:
-						extension.Log.Debugf("Received log event %v", logEvent.Type)
-						// Check the logEvent for runtimeDone and compare the RequestID
-						// to the id that came in via the Next API
-						if logEvent.Type == logsapi.RuntimeDone {
-							if logEvent.Record.RequestId == event.RequestID {
-								extension.Log.Info("Received runtimeDone event for this function invocation")
-								runtimeDoneSignal <- struct{}{}
-								return
-							} else {
-								extension.Log.Debug("Log API runtimeDone event request id didn't match")
-							}
-						}
-					}
-				}
-			}()
+			extension.StartBackgroundApmDataForwarding(apmServerTransport, funcDone, &backgroundDataSendWg)
```

Comment:
Can you please remind me why we only forward data to APM Server while the invocation is active? Why don't we start forwarding the data in the background outside the loop?

Is it to do with flushing at the end of the invocation? If that is the case, then perhaps we could find an alternative means of flushing, e.g. call StartBackgroundApmDataForwarding once before the loop, and allow it to receive requests over a channel to flush.

Reply:
Lambda users are billed based on the total execution time of their function - this includes extension execution time. To quote the [docs](https://docs.aws.amazon.com/lambda/latest/dg/runtimes-extensions-api.html):

> Extensions can impact function performance. As an extension author, you have control over the performance impact of your extension. For example, if your extension performs compute-intensive operations, the function's duration increases because the extension and the function code share the same CPU resources. In addition, if your extension performs extensive operations after the function invocation completes, the function duration increases because the Invoke phase continues until all extensions signal that they are completed.

However, your suggested implementation makes sense - if the Lambda environment is able to freeze the main for loop, why would we need to tie the data-forwarding goroutine to a single event?

However, after running a few tests, while getting the goroutines outside of the main for loop allows for successful APM data chunks forwarding, the goroutine is still running by the time we reach the end of the current invocation and will prevent the main thread from being frozen at the end of the active invocation, leading to the Lambda function execution reaching the defined timeout.